pkg/api/grpcServer.go

```diff
@@ -14,314 +14,204 @@
+		wordMap:          wordMap,
 	}
 }
 
-// dispatch does all the nitty-gritty details of handling Mixer's low-level API
-// protocol and dispatching to the right API dispatchWrapperFn.
-func (s *grpcServer) dispatch(stream grpc.Stream, methodName string, getState stateGetterFn, worker dispatchFn) error {
-
-	// tracks attribute state for this stream
-	reqTracker := s.attrMgr.NewTracker()
-	respTracker := s.attrMgr.NewTracker()
-	defer reqTracker.Done()
-	defer respTracker.Done()
-
-	// used to serialize sending on the grpc stream, since the grpc stream is not multithread-safe
-	sendLock := &sync.Mutex{}
-
-	root, ctx := s.tracer.StartRootSpan(stream.Context(), methodName)
-	defer root.Finish()
-
-	// ensure pending stuff is done before leaving
-	wg := sync.WaitGroup{}
-	defer wg.Wait()
-
-	for {
-		dState := getState()
-
-		// get a single message
-		err := stream.RecvMsg(dState.request)
-		if err == io.EOF {
-			return nil
-		} else if err != nil {
-			glog.Errorf("Reading from gRPC request stream failed: %v", err)
-			return err
-		}
-		// ensure the indices are matched for processing
-		*dState.responseIndex = *dState.requestIndex
-
-		requestBag, err := reqTracker.ApplyProto(dState.inAttrs)
-		if err != nil {
-			msg := "Request could not be processed due to invalid 'attribute_update'."
-			glog.Error(msg, "\n", err)
-			details := status.NewBadRequest("attribute_update", err)
-			*dState.result = status.InvalidWithDetails(msg, details)
-
-			sendLock.Lock()
-			err = s.sendMsg(stream, dState.response)
-			sendLock.Unlock()
-
-			if err != nil {
-				glog.Errorf("Unable to send gRPC response message: %v", err)
-			}
-
-			continue
-		}
+// dispatch implements all the nitty-gritty details of handling  Mixer's low-level API
+// protocol and dispatching to an appropriate API worker.
+func (s *grpcServer) dispatch(ctx context.Context, dState *dispatchState, worker dispatchFn) error {
+	requestBag, err := attribute.GetBagFromProto(dState.inAttrs, s.words)
+	if err != nil {
+		msg := "Request could not be processed due to invalid 'attributes'."
+		glog.Error(msg, "\n", err)
+		details := status.NewBadRequest("attributes", err)
+		out := status.InvalidWithDetails(msg, details)
+		return makeGRPCError(out)
+	}
 
-		// throw the message into the work queue
-		wg.Add(1)
-		s.gp.ScheduleWork(func() {
-			span, ctx2 := s.tracer.StartSpanFromContext(ctx, "RequestProcessing")
-			span.LogFields(log.Object("gRPC request", dState.request))
+	responseBag := attribute.GetMutableBag(nil)
 
-			responseBag := attribute.GetMutableBag(nil)
+	out := s.aspectDispatcher.Preprocess(ctx, requestBag, responseBag)
+	if status.IsOK(out) {
 
+		if err := requestBag.Merge(responseBag); err != nil {
+			// TODO: better error messages that push internal details into debuginfo messages
+			glog.Errorf("Could not merge mutable bags for request: %v", err)
+			out = status.WithInternal("The results from the request preprocessing could not be merged.")
+		} else {
```

Comment:
there is nothing wrong per-se. it obviously works. this is more about readability.

See: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Dealing with the error cases, etc., up front and indenting only for them makes it easier (imho) to quickly grasp the happy path. This is why I was suggesting check for !OK first to start with.

I don't want this to become a blocker for this PR, but I think it would improve the code. Up to your discretion.